### ISSUE TITLE
Concurrent "docker build" takes longer than sequential builds

### ISSUE THREAD

**Issue by dnephin** [moby/moby#9656]
### Background

We have a few build servers that run many builds in parallel. When a "docker build" step is added to the build,  it starts taking an order of magnitude longer to run (from 20min to hours). We've observed this with both docker 1.1.2 and 1.3.1. 

The builds do roughly this:

```
docker pull
docker build  # build multiple images
docker run    # run each of the built images
...
```
### The Issue

I tried to create a minimal reproduction of the issue at https://github.com/dnephin/dockerd-load-test

What I observed was that concurrent builds actually take longer to run than if I had run them sequentially. Tested with docker 1.3.3 and 1.4.0.

```
build 1: elapsed 8.411
build 2: elapsed 19.825
build 3: elapsed 30.438
build 4: elapsed 41.650
build 5: elapsed 54.865
build 10: elapsed 115.551
build 20: elapsed 321.406
```

As more concurrent builds are run, the elapsed time actually increases beyond n \* 8.4s.

See the [the repo](https://github.com/dnephin/dockerd-load-test) for the code to reproduce.

**Comment by anandkumarpatel** [comment]
I have seen the same Issue. build times seem to grow polynomially
few questions: 
Could we technically remove the lock if we guarantee we are dealing with different images per docker thread?
Do all the backends have this limitation? (overlay, btrfs)

the data:
the dockerfile and scripts I ran can be found here: https://github.com/anandkumarpatel/node-hello-world

here is a graph of parallel builds (1 to 261)
![time vs p builds](https://cloud.githubusercontent.com/assets/2194285/5599699/aca23564-9280-11e4-8f43-4ae6272343eb.png)

here is graph 261 builds ran one after the other
![time vs build number](https://cloud.githubusercontent.com/assets/2194285/5599668/3983ba94-9280-11e4-9bb1-89d262695a86.png)

**Comment by LK4D4** [comment]
I see that overlay has lock too. I am not sure that problem in that lock, though.

**Comment by cnelson** [comment]
I'm also experiencing this behavior.  Even running two builds in parallel kills performance.  I'm on btrfs.

`
docker version && docker info

Client version: 1.4.1
Client API version: 1.16
Go version (client): go1.3.3
Git commit (client): 5bc2ff8-dirty
OS/Arch (client): linux/amd64
Server version: 1.4.1
Server API version: 1.16
Go version (server): go1.3.3
Git commit (server): 5bc2ff8-dirty
Containers: 13
Images: 227
Storage Driver: btrfs
 Build Version: Btrfs v3.17.1
 Library Version: 101
Execution Driver: native-0.2
Kernel Version: 3.18.1
Operating System: CoreOS 557.2.0
CPUs: 9
Total Memory: 22.44 GiB
`

**Comment by boucher** [comment]
I believe I'm running into this issue, I'm not sure if its still on anyone else's radar.

**Comment by willejs** [comment]
+1

**Comment by jeffb-stell** [comment]
+1

**Comment by aelsabbahy** [comment]
+1

**Comment by phemmer** [comment]
To quote @GordonTheTurtle 

> Please don't use "+1" or "I have this too" comments on issues.

The notification spam is annoying.
Also, github has a feature specifically for this: https://github.com/blog/2119-add-reactions-to-pull-requests-issues-and-comments
Thank you, have a wonderful day.

**Comment by rajiff** [comment]
@anandkumarpatel  How did that graph got generated, which tool used

**Comment by anandkumarpatel** [comment]
@rajiff look at the repo I reference. In particular the parallel script. It prints out the times and I paste that into Google sheets and made the graph.

the data:
the dockerfile and scripts I ran can be found here: https://github.com/anandkumarpatel/node-hello-world

**Comment by bemeyert** [comment]
Ping. Any news on this one?

**Comment by jtaylor100** [comment]
I am looking to optimise our current deployment process and this is currently a bottleneck.

**Comment by thaJeztah** [comment]
Giving this another spin;

```
Client:
 Version:           18.06.0-ce
 API version:       1.38
 Go version:        go1.10.3
 Git commit:        0ffa825
 Built:             Wed Jul 18 19:09:54 2018
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.06.0-ce
  API version:      1.38 (minimum version 1.12)
  Go version:       go1.10.3
  Git commit:       0ffa825
  Built:            Wed Jul 18 19:07:56 2018
  OS/Arch:          linux/amd64
  Experimental:     true


Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 2
Server Version: 18.06.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: d64c661f1d51c48782c9cec8fda7604785f93587
runc version: 69663f0bd4b60df09991c08812a60108003fa340
init version: fec3683
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.15.0-29-generic
Operating System: Ubuntu 18.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.852GiB
Name: ubuntu-s-2vcpu-4gb-ams3-01
ID: CDEY:Z5NG:HV33:J33H:IWHT:B46I:MT4P:L7Q4:QAFN:36V2:34QR:R6Y3
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 22
 Goroutines: 43
 System Time: 2018-07-31T23:52:56.274093549Z
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Labels:
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

`DOCKER_BUILDKIT=1`

```
INFO:root:build 1:   elapsed  0.202 = 0.2020 per build
INFO:root:build 5:   elapsed  0.415 = 0.0830 per build
INFO:root:build 10:  elapsed  0.875 = 0.0875 per build
INFO:root:build 50:  elapsed  4.080 = 0.0816 per build
INFO:root:build 100: elapsed  8.723 = 0.0870 per build
INFO:root:build 200: elapsed 16.155 = 0.0807 per build
```

`DOCKER_BUILDKIT=0`

```
INFO:root:build 1:   elapsed  0.602 = 0.6020 per build
INFO:root:build 5:   elapsed  1.240 = 0.2480 per build
INFO:root:build 10:  elapsed  1.863 = 0.1863 per build
INFO:root:build 50:  elapsed  4.604 = 0.0921 per build
INFO:root:build 100: elapsed  9.495 = 0.0950 per build
INFO:root:build 200: elapsed 18.864 = 0.0943 per build
```

**Comment by jomach** [comment]
I have the same issue, on my side is the pushing to image to the registry that seems extreme slow when multiple pushes are occuring. 

`Client:
 Version:      17.09.1-ce
 API version:  1.32
 Go version:   go1.8.5
 Git commit:   f4ffd2511ce9
 Built:        Wed Feb 28 13:09:32 2018
 OS/Arch:      linux/amd64

Server:
 Version:      17.09.1-ce
 API version:  1.32 (minimum version 1.12)
 Go version:   go1.8.5
 Git commit:   f4ffd2511ce9
 Built:        Sat Oct 21 17:27:07 2017
 OS/Arch:      linux/amd64
 Experimental: false`

**Comment by thaJeztah** [comment]
@jomach Docker 17.09 is quite old, and reached EOL in December 2017; are you still seeing the issue in a current version of Docker?

**Comment by jomach** [comment]
@thaJeztah  I just want to notice that the issue happens on a older version to. Yes we need to upgrade...

**Comment by marcellodesales** [comment]
@thaJeztah what's the diff between this support and docker-compose? https://github.com/docker/compose/pull/6115 It seems to be working just fine over there...

**Comment by thehappycoder** [comment]
How do you actually run two builds of different images in parallel? I am using `18.06.1-ce-mac73 (26764)`

**Comment by Stqs** [comment]
Hi All,

any chance it will be solved soon?
what are current workarounds options?
can somebody explain technically why exactly this happens?

Thanks

**Comment by thaJeztah** [comment]
@Stqs in my latest testing (see https://github.com/moby/moby/issues/9656#issuecomment-409404398) the problem no longer reproduced, but perhaps the reproducer is not covering all cases; are you able to provide a reproducer to verify the issue?

**Comment by tonistiigi** [comment]
Closing this. BuildKit is highly optimized for parallel workloads. If you see anything like this in buildkit or buildkit compared to legacy builder please report a new issue with a repro case.
